checker/src/org/checkerframework/checker/index/upperbound/UpperBoundVisitor.java

```diff
@@ -36,6 +55,129 @@ public UpperBoundVisitor(BaseTypeChecker checker) {
     }
 
     /**
+     * This visits all local variable declarations and issues a warning if a dependent annotation is
+     * written on a local variable. This is necessary for the soundness of the reassignment code,
+     * which must unrefine all qualifiers, but which cannot collect all in-scope local variables. So
+     * that there are no local variables with qualifiers that are not in the store, this method
+     * forbids programmers from writing such qualifiers. This may introduce false positives if the
+     * programmer would like to write a non-primary annotation on a local variable.
+     *
+     * <p>Also issues warnings if the expression in a dependent annotation is not permitted.
+     * Permitted expressions in dependent annotations are:
+     *
+     * <ul>
+     *   <li>final or effectively final variables
```

Comment:
No. This is any final or effectively final variable, in any scope. Because such variables can't be reassigned, there's no need to prevent users from writing annotations that are dependent on them: they can never be invalidated by a side-effect, since no side-effect can reassign them!

Reply:
I mean what kinds of variables have to be final or effectively final? Local variables and private fields are allowed even if they are not final or effectively final.  So which variables have to be final or effectively final?  Non-private fields and methods parameters?

Reply:
I see the confusion. If a variable has any one of the properties on this list, it is permitted. So, being final or effectively final is sufficient by itself; so is being a private field; so is being a local variable; etc. I can rephrase this comment to make this clear.